pennylane/optimize/spsa.py

```diff
@@ -0,0 +1,255 @@
+        <https://www.jhuapl.edu/SPSA/PDF-SPSA/Spall_An_Overview.PDF>`_
+
+    .. note::
+
+        * The number of quantum device executions is :math:`2*iter*num terms hamiltonian`.
+        * In case of using ``step_and_cost`` method instead of ``step``, the number
+          of executions increment to calculate the cost function.
+
+    .. note::
+
+        In cases of hybrid classical-quantum workflows:
+
+        * In implementation of a QNode as a layer of a Keras sequential model,
+          possible optimizers for the model are from the classical platform i.e.
+          `tf.keras.optimizers.SGD
+          <https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/SGD>`_.
+        * In a hybrid classical-quantum-classical workflow where we use SPSAOptimizer
+          for the quantum part, we have to extract the values of the classical tensor
+          in order to use it in the quantum circuit as inputs and assign the output
+          of the quantum circuit to a subsecuent classical tensor.
+
+
+    **Examples:**
+
+    For VQE/VQE-like problems, the objective function can be the following:
+
+    >>> dev = qml.device("default.qubit", wires=1)
+    >>> def circuit(params, wires):
+    ...    qml.BasisState(np.array([1, 1, 0, 0]), wires=wires)
+    ...    for i in wires:
+    ...        qml.Rot(*params[i], wires=i)
+    ...    qml.CNOT(wires=[2, 3])
+    ...    qml.CNOT(wires=[2, 0])
+    ...    qml.CNOT(wires=[3, 1])
+    >>> def exp_val_circuit(params):
+    ...    circuit(params, range(dev.num_wires))
+    ...    return qml.expval(h2_ham)
+    >>> params = np.random.normal(0, np.pi, (num_qubits, 3), requires_grad=True)
+    >>> cost = qml.QNode(exp_val_circuit, dev)
+
+    Once constructed, the cost function can be passed directly to the
+    optimizer's ``step`` or ``step_and_cost`` function:
+
+    >>> max_iterations = 100
+    >>> opt = qml.SPSAOptimizer(maxiter=max_iterations)
+    >>> for n in range(max_iterations):
+    ...     params, energy = opt.step_and_cost(cost, params)
+
+    Example of hybrid classical-quantum workflow:
+
+    >>> dev = qml.device("default.qubit", wires=n_qubits)
+    >>> @qml.qnode(dev, interface=None)
+    >>> def layer_fn_spsa(inputs, weights):
+    ...     qml.AngleEmbedding(inputs, wires=range(n_qubits))
+    ...     qml.BasicEntanglerLayers(weights, wires=range(n_qubits))
+    ...     return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_qubits)]
+
+    >>> opt = qml.SPSAOptimizer(maxiter=max_iterations)
+    >>> tensor_in = tf.Variable([0.27507603, 0.3453423])
+    >>> params = tf.Variable([[3.97507603, 2.00854038],
+    ...                       [3.12950603, 3.00854038],
+    ...                       [1.17907603, 1.10854038],
+    ...                       [0.97507603, 1.00854038],
+    ...                       [1.25907603, 0.40854088]])
+    >>> tensor_out = tf.Variable([0,0])
+
+    >>> init = tf.compat.v1.global_variables_initializer()
+    >>> with tf.compat.v1.Session() as sess:
```

Comment:
Note two things here:
* Usually PennyLane optimizers (such as `qml.SPSAOptimizer`) can not be used with other interfaces ([see here](https://pennylane--2661.org.readthedocs.build/en/2661/introduction/interfaces/tf.html#optimization-using-tensorflow)) - here it might work well because we SPSA uses the results of cost function evaluations;
* We would like to refrain from using TensorFlow v1 as PennyLane only support TF v2.